### PR TITLE
Add CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by arnout**
Make a start on a CONTRIBUTING file.

### CODE REVIEW

**Review by simonbaren — APPROVED** [review]

**Review by simonbaren — APPROVED** [review]

**Comment by arnout** [comment]
@simonbaren This is still [RFC] so it's not the intention that you approve it :-)

**Review by tomereli — APPROVED** [review]

**Comment by tomereli** [comment]
Ah RFC = Request for comments🤣

File: CONTRIBUTING.md
Comment (tomereli) [inline CONTRIBUTING.md:27]: I think we can add a sub section here stating that the documentation we now have in .ppt, .docx and .pdf formats should be ported to .md. Once this is done we can point there from our wiki.

File: CONTRIBUTING.md
Comment (tomereli) [inline CONTRIBUTING.md:45]: I assume you plan to update these sections before we merge this.
I think that the commits and pull requests sections should have a link to an "example" commit, we have plenty of those.
Maybe also add a couple of words explaining why it is important to have small well-documented commits, and why it is bad not to.

In addition, you should explain how a developer should start working even before asking for code review - using a WIP branch and committing all the way. Finally, when the change is ready, open feature branches for each "logical change" and start cherry-picking, fixing commit messages as you go along.

Review fixups - this should be explained here as well, in detail - we want contributors (and by contributors I mean us maintainers as well) to use interactive rebase and fixup commits. For this it is crucial to have well documented and logical commits otherwise finding the commit we want to fixup is a nightmare.

This gets me to the next section I think should be here - Code Review. We want contributors to participate, so some best code reviewing practices can be helpful. Especially regarding reviewing commits instead of pull requests, so that the fixups will go smoother (I'm not sure I know till now if that works:) )

Definition of done - I know we should "soon" have CI with testing capabilities but until we do, we should mention here what is the definition of done (DoD), and that without it we won't review pull requests.
DoD should for now include building and testing with ZMQ. Until we have easy to use containers instructions for running 2 or more nodes, we should ask the following:
1. Build with `./maptools.py build all -f BUILD_TESTS=ON MSGLIB=zmq BWL_TYPE=DWPAL -G Ninja`
2. Test your changes - run the local_bus, ieee1905_transport, agent and controller. Tested means that you saw OPERATIONAL in the agent log.
3. Run clang-format.sh. If it fixes lines you did not change, commit that separately.
4. Make sure no new warnings are added
5. optional - valgrind / klockwork / clang-tidy / etc - we should add this so we don't forget, and say that currently this can be skipped.

File: CONTRIBUTING.md
Comment (tomereli) [inline CONTRIBUTING.md:41]: We should add "Development workflow" section, which points to the README.md file so that the potential contributor can easily know where to start. Also maybe couple of words on how to where to start - reading the docs, watching the Videos from the workshop (I guess they will be uploaded?), running controller & agent and sniffing packets, etc.

Under this section will be the Commits, Pull Requests and Coding style sections you added.

**Review by tomereli — CHANGES_REQUESTED** [review]
Some sections we should consider adding:
1. Code of conduct - I see that in every CONTRIBUTERS.md file (sometimes it is a separate file).  [This](https://gist.github.com/PurpleBooth/b24679402957c63ec426) and [Facebook's](https://code.fb.com/codeofconduct/) can be a good place to start.

2. branch names - we should select branch names and ask contributors to follow this. I think that features should be feature/<> or dev/<> (you choose), bugfixes should be bugfix/<>, small fixes should be hotfix/<>, etc. 
Maybe that's only what I'm used to so feel free to choose different convention.

3. Versioning - we already decided on versioning so maybe add this also?

4. Bugs - We should mention that all bugs should be documented in github issues, and define tags to be used (maybe in a later stage though). Some contributors might be contributing by finding bugs which is welcome, so we should state that.

**Comment by arnout** [comment]
> Some sections we should consider adding:
> 
>     1. Code of conduct - I see that in every CONTRIBUTERS.md file (sometimes it is a separate file).  [This](https://gist.github.com/PurpleBooth/b24679402957c63ec426) and [Facebook's](https://code.fb.com/codeofconduct/) can be a good place to start.

I think https://prplfoundation.org/about/code-of-conduct/ is a better one :-)



>     2. branch names - we should select branch names and ask contributors to follow this. I think that features should be feature/<> or dev/<> (you choose), bugfixes should be bugfix/<>, small fixes should be hotfix/<>, etc.
>        Maybe that's only what I'm used to so feel free to choose different convention.

Completely agree that we should define a naming convention. I don't care much what it is, as long as there is something. Note, however, that external contributors will not create branches in the prplfoundation repo so it doesn't really belong to CONTRIBUTING.md.


>     3. Versioning - we already decided on versioning so maybe add this also?

 I was planning to add everything we discussed last Monday somewhere in the docs.


>     4. Bugs - We should mention that all bugs should be documented in github issues, and define tags to be used (maybe in a later stage though). Some contributors might be contributing by finding bugs which is welcome, so we should state that.

Have you looked at the second bullet? :-)

File: CONTRIBUTING.md
Comment (arnout) [reply CONTRIBUTING.md:45]: > I think that the commits and pull requests sections should have a link to an "example" commit, we have plenty of those.

Good idea. I couldn't really find the ideal example however. Let's add that later.

File: CONTRIBUTING.md
Comment (arnout) [reply CONTRIBUTING.md:41]: I don't know what exactly you meant with this.

I propose that you just commit what I will push soon, and make corrections to it yourself.

**Review by tomereli — APPROVED** [review]
Other comments will be fixed in later changes

**Pull request merged**
